src/ts/deploy.ts

```diff
@@ -0,0 +1,88 @@
+import { utils } from "ethers";
+import { ethers } from "hardhat";
+import { DeployResult, DeploymentsExtension } from "hardhat-deploy/types";
+
+export const contractNames = {
+  authenticator: "GPv2AllowListAuthentication",
+  settlement: "GPv2Settlement",
+};
+
+/**
+ * The salt used when deterministically deploying smart contracts.
+ */
+export const salt = utils.formatBytes32String("dev");
+
+/**
+ * The contract used to deploy contracts deterministically with CREATE2.
+ * The address is chosen by the hardhat-deploy library.
+ * It is the same in any EVM-based network.
+ *
+ * https://github.com/Arachnid/deterministic-deployment-proxy
+ */
+const DEPLOYER_CONTRACT = "0x4e59b44847b379578588920ca78fbf26c0b4956c";
+
+/**
+ * Computes the deterministic address at which the contract will be deployed.
+ * This address does not depend on which network the contract is deployed to.
+ *
+ * @param contractName Name of the contract for which to find the address.
+ * @param deploymentArguments Extra arguments that are necessary to deploy.
+ * @returns The address that is expected to store the deployed code.
+ */
+export async function deterministicDeploymentAddress(
+  contractName: string,
+  ...deploymentArguments: unknown[]
+): Promise<string> {
+  const factory = await ethers.getContractFactory(contractName);
+  const deployTransaction = factory.getDeployTransaction(
+    ...deploymentArguments,
+  );
+
+  return utils.getCreate2Address(
+    DEPLOYER_CONTRACT,
+    salt,
+    utils.keccak256(deployTransaction.data || "0x"),
+  );
+}
+
+/**
+ * Print to screen the result of a successful contract deployment.
+ *
+ * @param deployResult Result of the deployment.
+ * @param contractName Name of the deployed contract.
+ * @param networkName Name of the network to which the contract is deployed.
+ * @param log The logging function.
+ */
+export async function logResult(
```

Comment:
This function is neat, though I would simply use the existing logging function. You can just do:

```
const deployResult = await deploy(settlement, {
    from: deployer,
    gasLimit: 2000000,
    args: [authenticatorAddress],
    deterministicDeployment: salt,
    log: true,
  });
```
Then we have less code to maintain

Reply:
I implemented the logging function because I noticed that the examples in the `truffle-deploy` documentation used custom logging. You are right that it could be removed.

For comparison, this is the output with the standard logging with a real deployment:
```
sending eth to create2 contract deployer address (0x3fab184622dc19b6109349b94811493bf2a45362)...
deploying create2 deployer contract (at 0x4e59b44847b379578588920ca78fbf26c0b4956c) using deterministic deployment (https://github.com/Arachnid/deterministic-deployment-proxy)...
"GPv2AllowListAuthentication" deployed at 0x152175a095A26238A18c2B7186DD5Beb964Fc4a4 with 570422 gas
"GPv2Settlement" deployed at 0x0275fAd9FCDbB62fc9D4555D2a6BC0e367085f92 with 787322 gas
Done in 4.15s.
```
And this is what it looks like when the contract was already deployed:
```
reusing "GPv2AllowListAuthentication" at 0x152175a095A26238A18c2B7186DD5Beb964Fc4a4
reusing "GPv2Settlement" at 0x0275fAd9FCDbB62fc9D4555D2a6BC0e367085f92
Done in 3.40s.
```

I like the extra info in the custom logging function because it helps double checking the input parameters (gas price and network), but I see your point. I'll leave this up for discussion and remove the code if there's no interest in this.